flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManagerTest.java

```diff
@@ -0,0 +1,1092 @@
+
+			// a second request is only sent if the first request timed out
+			secondSlotRequestFuture.get();
+		}
+	}
+
+	/**
+	 * Tests that a slot allocation is retried if it times out on the task manager side.
+	 */
+	@Test
+	public void testTaskManagerSlotAllocationTimeoutHandling() throws Exception {
+		final JobID jobId = new JobID();
+		final ResourceRequirements resourceRequirements = createResourceRequirementsForSingleSlot(jobId);
+		final CompletableFuture<Acknowledge> slotRequestFuture1 = new CompletableFuture<>();
+		final CompletableFuture<Acknowledge> slotRequestFuture2 = new CompletableFuture<>();
+		final Iterator<CompletableFuture<Acknowledge>> slotRequestFutureIterator = Arrays.asList(slotRequestFuture1, slotRequestFuture2).iterator();
+		final ArrayBlockingQueue<SlotID> slotIds = new ArrayBlockingQueue<>(2);
+
+		final TestingTaskExecutorGateway taskExecutorGateway = new TestingTaskExecutorGatewayBuilder()
+			.setRequestSlotFunction(FunctionUtils.uncheckedFunction(
+				requestSlotParameters -> {
+					slotIds.put(requestSlotParameters.f0);
+					return slotRequestFutureIterator.next();
+				}))
+			.createTestingTaskExecutorGateway();
+
+		final ResourceID resourceId = ResourceID.generate();
+		final TaskExecutorConnection taskManagerConnection = new TaskExecutorConnection(resourceId, taskExecutorGateway);
+
+		final SlotID slotId1 = new SlotID(resourceId, 0);
+		final SlotID slotId2 = new SlotID(resourceId, 1);
+		final SlotReport slotReport = new SlotReport(Arrays.asList(createFreeSlotStatus(slotId1), createFreeSlotStatus(slotId2)));
+
+		final ResourceTracker resourceTracker = new DefaultResourceTracker();
+		final DefaultSlotTracker slotTracker = new DefaultSlotTracker();
+
+		try (DeclarativeSlotManager slotManager = createDeclarativeSlotManagerBuilder()
+			.setResourceTracker(resourceTracker)
+			.setSlotTracker(slotTracker)
+			.buildAndStartWithDirectExec()) {
+
+			slotManager.registerTaskManager(taskManagerConnection, slotReport);
+
+			slotManager.processResourceRequirements(resourceRequirements);
+
+			final SlotID firstSlotId = slotIds.take();
+			assertThat(slotIds, is(empty()));
+
+			DeclarativeTaskManagerSlot failedSlot = slotTracker.getSlot(firstSlotId);
+
+			// let the first attempt fail --> this should trigger a second attempt
+			slotRequestFuture1.completeExceptionally(new SlotAllocationException("Test exception."));
+
+			assertThat(getTotalResourceCount(resourceTracker.getAcquiredResources(jobId)), is(1));
+
+			// the second attempt succeeds
+			slotRequestFuture2.complete(Acknowledge.get());
+
+			final SlotID secondSlotId = slotIds.take();
+			assertThat(slotIds, is(empty()));
+
+			DeclarativeTaskManagerSlot slot = slotTracker.getSlot(secondSlotId);
+
+			assertThat(slot.getState(), is(SlotState.ALLOCATED));
+			assertEquals(jobId, slot.getJobId());
+
+			if (!failedSlot.getSlotId().equals(slot.getSlotId())) {
+				assertThat(failedSlot.getState(), is(SlotState.FREE));
+			}
```

Comment:
It can happen that we retry with `firstSlotId`?

Reply:
yes; when the allocation fails we go through the same process as on the first attempt, and neither the `SlotTracker#getFreeSlots` nor `SlotMatchingStrategy#findMatchingSlot` give any guarantees on the order in which slots are returned/matched.